Bring back and finish image loading optimization

alpaix opened the pull request:
This change is to bring back image loading optimization while revisiting its implementation.

Key changes are:
- Use of MemoryCache instead of a regular dictionary.
- Image loading logic moved from a model to a value convertor.
- Use of shared style for package icon images.
- Code factorization to allow further re-use.
- Sliding error threshold

Fixes https://github.com/NuGet/Home/issues/2187.

//cc @yishaigalatzer @emgarten @rrelyea

rrelyea commented on src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs:
```diff
@@ -0,0 +1,70 @@
+    internal class ErrorFloodGate
+    {
+        // If we fail at least this high (failures/attempts), we'll shut off image loads.
+        // TODO: Should we allow this to be overridden in nuget.config.
```
lets decide if this is necessary or not...and kill the todo.

rrelyea commented on src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs:
```diff
@@ -0,0 +1,70 @@
+
+        private static int GetTicks(DateTimeOffset origin)
+        {
+            return (int)((DateTimeOffset.Now - origin).TotalSeconds / 5);
```
never suspected that a tick was 5 seconds. made the code hard to interpret in IsOpen.
combination of comments and/or method name renaming might be good.

alpaix replied:
You're right. Will add more comments on that.

rrelyea commented on src/NuGet.Clients/PackageManagement.UI/Resources/Resources.xaml:
```diff
@@ -135,27 +135,23 @@
+          ConverterParameter="{x:Static nuget:Images.DefaultPackageIcon}"
+          TargetNullValue="{x:Static nuget:Images.DefaultPackageIcon}"
+          FallbackValue="{x:Static nuget:Images.DefaultPackageIcon}"
+          IsAsync="False"/>
```
IsAsync false is default behavior. better to not set values to default values in xaml.

alpaix replied:
Agree

rrelyea commented:
:shipit:

alpaix committed:
Image loading optimization of InfiniteScrollList

Rework of 6d112fb.
Fixes https://github.com/NuGet/Home/issues/2154.

Key changes:
- Use of MemoryCache instead of regular dictionary.
- Image loading logic moved from a model to a value convertor.
- Use of shared style for package icon images.
- Code factorization to allow further re-use.
- Sliding error threshold
